File: backend/app/schemas/meta.py
from pydantic import BaseModel, Field


class CategoryIn(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    display_name: str | None = Field(default=None, max_length=64)
    color: str | None = Field(default=None, max_length=16)
    sort_order: int = 0


class CategoryPatchIn(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=64)
    display_name: str | None = Field(default=None, max_length=64)
    color: str | None = Field(default=None, max_length=16)
    sort_order: int | None = None


class CategoryOut(BaseModel):
    id: int
    name: str
    display_name: str | None
    color: str | None
    sort_order: int

    model_config = {"from_attributes": True}


class CategoryReorderIn(BaseModel):
    order: list[dict]    # [{id: int, sort_order: int}]


class TagIn(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    color: str | None = Field(default=None, max_length=16)


class TagOut(BaseModel):
    id: int
    name: str
    color: str | None

    model_config = {"from_attributes": True}


class DictIn(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    target_field: str = Field(default="any", pattern="^(payee|item_name|any)$")
    remark: str | None = Field(default=None, max_length=255)


class DictPatchIn(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=64)
    target_field: str | None = Field(default=None, pattern="^(payee|item_name|any)$")
    remark: str | None = Field(default=None, max_length=255)


class DictOut(BaseModel):
    id: int
    name: str
    target_field: str
    remark: str | None
    entry_count: int = 0

    model_config = {"from_attributes": True}


class DictEntryIn(BaseModel):
    key_text: str = Field(min_length=1, max_length=255)
    category_id: int


class DictEntryOut(BaseModel):
    id: int
    key_text: str
    category_id: int

    model_config = {"from_attributes": True}


class DictBulkImportIn(BaseModel):
    csv_text: str        # 每行 "key,category_name"，忽略空行与 # 注释

File: frontend/src/api/meta.ts
import request from './request'

export interface Category {
  id: number
  name: string
  display_name: string | null
  color: string | null
  sort_order: number
}

export interface Tag {
  id: number
  name: string
  color: string | null
}

export interface UserDict {
  id: number
  name: string
  target_field: 'payee' | 'item_name' | 'any'
  remark: string | null
  entry_count: number
}

export interface DictEntry {
  id: number
  key_text: string
  category_id: number
}

export const categoryApi = {
  list: () => request.get<unknown, Category[]>('/categories'),
  create: (body: Omit<Category, 'id'>) => request.post<unknown, Category>('/categories', body),
  patch: (id: number, body: Partial<Category>) => request.patch<unknown, Category>(`/categories/${id}`, body),
  remove: (id: number) => request.delete<unknown, { deleted: number }>(`/categories/${id}`),
  reorder: (order: Array<{ id: number; sort_order: number }>) =>
    request.post<unknown, { reordered: number }>('/categories/reorder', { order }),
}

export const tagApi = {
  list: () => request.get<unknown, Tag[]>('/tags'),
  create: (body: { name: string; color?: string | null }) => request.post<unknown, Tag>('/tags', body),
  remove: (id: number) => request.delete<unknown, { deleted: number }>(`/tags/${id}`),
}

export const dictApi = {
  list: () => request.get<unknown, UserDict[]>('/dicts'),
  create: (body: Omit<UserDict, 'id' | 'entry_count'>) => request.post<unknown, UserDict>('/dicts', body),
  patch: (id: number, body: Partial<UserDict>) => request.patch<unknown, UserDict>(`/dicts/${id}`, body),
  remove: (id: number) => request.delete<unknown, { deleted: number }>(`/dicts/${id}`),
  listEntries: (id: number) => request.get<unknown, DictEntry[]>(`/dicts/${id}/entries`),
  addEntry: (id: number, body: { key_text: string; category_id: number }) =>
    request.post<unknown, DictEntry>(`/dicts/${id}/entries`, body),
  removeEntry: (id: number, entryId: number) =>
    request.delete<unknown, { deleted: number }>(`/dicts/${id}/entries/${entryId}`),
  bulkImport: (id: number, csv_text: string) =>
    request.post<unknown, { inserted: number }>(`/dicts/${id}/entries/bulk`, { csv_text }),
}
